openedx/core/djangoapps/programs/utils.py

```diff
@@ -132,28 +133,142 @@ def get_display_category(program):
+        self.engaged_programs = self._find_engaged_programs(self.user)
+        self.course_certs = None
 
-    return engaged_programs
+    def _find_engaged_programs(self, user):
+        """Derive a list of programs in which the given user is engaged.
+
+        Arguments:
+            user (User): The user for which to find engaged programs.
+
+        Returns:
+            list of program dicts, ordered by most recent enrollment.
+        """
+        programs = get_programs(user)
+        flattened = flatten_programs(programs, self.course_ids)
+
+        engaged_programs = []
+        for course_id in self.course_ids:
+            for program in flattened.get(course_id, []):
+                if program not in engaged_programs:
+                    engaged_programs.append(program)
+
+        return engaged_programs
+
+    @property
+    def progress(self):
+        """Gauge a user's progress towards program completion.
+
+        Returns:
+            list of dict, each containing information about a user's progress
+                towards completing a program.
+        """
+        self.course_certs = get_completed_courses(self.user)
+
+        progress = []
+        for program in self.engaged_programs:
+            completed, in_progress, not_started = [], [], []
+
+            for course_code in program['course_codes']:
+                name = course_code['display_name']
+
+                if self._is_complete(course_code):
+                    completed.append(name)
+                elif self._is_in_progress(course_code):
+                    in_progress.append(name)
+                else:
+                    not_started.append(name)
+
+            progress.append({
+                'id': program['id'],
+                'completed': completed,
+                'in_progress': in_progress,
+                'not_started': not_started,
+            })
+
+        return progress
+
+    def _is_complete(self, course_code):
+        """Check if a user has completed a course code.
+
+        A course code qualifies as completed if the user has earned a
+        certificate in the right mode for any nested run.
+
+        Arguments:
+            course_code (dict): Containing nested run modes.
+
+        Returns:
+            bool, whether the course code is complete.
+        """
+        return any([
+            self._parse(run_mode) in self.course_certs
+            for run_mode in course_code['run_modes']
+        ])
+
+    def _is_in_progress(self, course_code):
+        """Check if a user is in the process of completing a course code.
+
+        A user is in the process of completing a course code if they're
+        enrolled in the course.
+
+        Arguments:
+            course_code (dict): Containing nested run modes.
+
+        Returns:
+            bool, whether the course code is in progress.
+        """
+        return any([
+            run_mode['course_key'] in self.course_ids
```

Comment:
In the `_is_complete` function above, the array returned are of format returned by `_parse`. But the format of the return elements here do not use `_parse` function. Why the different item structure? Also, I noticed for not started items, the element returned is a straight up name.